Add a CONTRIBUTING.md file (#161)
googleinterns/open-sesame: CONTRIBUTING.md

# Contributing guidelines

## Pull Request Checklist

Before sending your pull requests, make sure you followed this list.

> - Read [contributing guidelines](CONTRIBUTING.md).

Reviewer: This could be confusing. Maybe specify that the reader is already looking at CONTRIBUTING.md by just saying something like "Read this file completely."?

Author: Oh yeah I saw that, I put the wrong link in, my bad

> - Read [contributing guidelines](#contributing-code).

Author: ```suggestion
- Read [contributing guidelines](#contributing-guidelines-and-standards).
- Read our [README](README.md)
```

  * If you work for a company that wants to allow you to contribute your work, then you'll need to sign a [corporate CLA](https://code.google.com/legal/corporate-cla-v1.0.html).

Follow either of the two links above to access the appropriate CLA and instructions for how to sign and return it. Once we receive it, we'll be able to accept your pull requests.

***NOTE***: Only original source code from you and other people that have signed the CLA can be accepted into the main repository.

### Contributing code

If you have improvements to OpenSesame, send us your pull requests! For those
just getting started, Github has a
[how to](https://help.github.com/articles/using-pull-requests/).

OpenSesame team members will be assigned to review your pull requests. Once the
pull requests are approved by two team members and pass continuous integration
checks, an OpenSesame member will squash and merge your pul request with the 
master branch. After approval and merging, the site will be redeployed on a 

> quaterly basis. This is done to ensure that we get a good amount of progress 

Reviewer: Nit: quarterly
Also, out of curiosity, what made you land on quarterly updates?

Author: It felt like a good amount of time. I just looked into it and [this site](https://learn.g2.com/how-often-should-you-update-your-website) says it's good to update websites monthly, I guess I'll change it to monthly.
```suggestion
monthly basis. This is done to ensure that we get a good amount of progress 
```

onto the website before new versions are rolled out.

If you want to contribute, start working through the OpenSesame codebase,
navigate to the
[Github "issues" tab](https://github.com/googleinterns/open-sesame/issues) and start
looking through interesting issues. If you are not sure of where to start, then
start by trying one of the smaller/easier issues here i.e.
[issues with the "good first issue" label](https://github.com/googleinterns/open-sesame/labels/good%20first%20issue)
and then take a look at the
[issues with the "contributions welcome" label](https://github.com/googleinterns/open-sesame/labels/stat%3Acontributions%20welcome).
. If you want to get acquainted with the code first, It might be useful to
start by looking into our documentation and proposing/fixing some 
[issues with the "documentation" label](https://github.com/googleinterns/open-sesame/issues?q=is%3Aissue+label%3Adocumentation).
These are issues that we believe are particularly well suited for outside
contributions, often because we probably won't get to them right now and it is 

> easier to find holes in documentation if you are new to the codebase. 

Reviewer: Should we also encourage new contibutors to join Open Sesame and find a mentor for this project?

Author: good Idea!!!

***NOTE***: If you decide to start on an issue, leave a comment so that other people know that
you're working on it. If you want to help out, but not alone, use the issue
comment thread to coordinate.

### Contribution guidelines and standards

Before sending your pull request for
[review](https://github.com/googleinterns/open-sesame/pulls),
make sure your changes are consistent with the guidelines and follow the
OpenSesame coding style.

#### Coding style

* [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)
* [Google JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html)
* [Google HTML/CSS Style Guide](https://google.github.io/styleguide/htmlcssguide.html)

#### Running sanity check and tests

> ##### Running sanity check

Reviewer: Is it worth mentioning 'mvn clean' and 'npm install' here because they have helped us solve many otherwise mysterious issues?

Author: good point! It skipped my mind. It would make sense to have a common troubleshooting section

Author: should this be here or in the `README`?

Author: added it to the README

Please fill out either the individual or corporate Contributor License Agreement (CLA).

  * If you are an individual writing original source code and you're sure you own the intellectual property, then you'll need to sign an [individual CLA](https://code.google.com/legal/individual-cla-v1.0.html).
  * If you work for a company that wants to allow you to contribute your work, then you'll need to sign a [corporate CLA](https://code.google.com/legal/corporate-cla-v1.0.html).

Follow either of the two links above to access the appropriate CLA and instructions for how to sign and return it. Once we receive it, we'll be able to accept your pull requests.

***NOTE***: Only original source code from you and other people that have signed the CLA can be accepted into the main repository.

### Contributing code

If you have improvements to OpenSesame, send us your pull requests! For those
just getting started, Github has a
[how to](https://help.github.com/articles/using-pull-requests/).

OpenSesame team members will be assigned to review your pull requests. Once the
pull requests are approved by two team members and pass continuous integration

> checks, an OpenSesame member will squash and merge your pul request with the 

Author: ```suggestion
checks, an OpenSesame member will squash and merge your pull request with the 
```

master branch. After approval and merging, the site will be redeployed on a 
monthly basis. This is done to ensure that we get a good amount of progress 
onto the website before new versions are rolled out.

If you want to contribute, start working through the OpenSesame codebase,
navigate to the
[Github "issues" tab](https://github.com/googleinterns/open-sesame/issues) and start
looking through interesting issues. If you are not sure of where to start, then
start by trying one of the smaller/easier issues here i.e.
[issues with the "good first issue" label](https://github.com/googleinterns/open-sesame/labels/good%20first%20issue)
and then take a look at the
[issues with the "contributions welcome" label](https://github.com/googleinterns/open-sesame/labels/stat%3Acontributions%20welcome).

> . If you want to get acquainted with the code first, It might be useful to

Author: ```suggestion
 If you want to get acquainted with the code first, It might be useful to
```

# Contributing guidelines

## Pull Request Checklist

Before sending your pull requests, make sure you followed this list.

- Read [contributing guidelines](#contributing-guidelines-and-standards).
- Read our [README](README.md).
- Ensure you have signed the [Contributor License Agreement (CLA)](https://cla.developers.google.com/).
- Changes are consistent with the [Coding Style](#coding-style).

## How to become a contributor and submit your own code

### Contributor License Agreements

> We'd love to accept your patches! Before we can take them, we have to jump a couple of legal hurdles.

Author: That's a really good point!
```suggestion
We'd love to accept your patches! Before we can take them, you need to certify yourself as a Google Contributor. This allows the you to retain ownership of the code you submit while granting Google the necessary legal rights to use that contribution. The CLA only needs to be signed once and it covers all Google projects.
```

  * If you work for a company that wants to allow you to contribute your work, then you'll need to sign a [corporate CLA](https://code.google.com/legal/corporate-cla-v1.0.html).

Follow either of the two links above to access the appropriate CLA and instructions for how to sign and return it. Once we receive it, we'll be able to accept your pull requests.

***NOTE***: Only original source code from you and other people that have signed the CLA can be accepted into the main repository.

### Contributing code

***NOTE***: We **highly** recommend using [OpenSesame](https://step2020-279820.uc.r.appspot.com/)
 to find a mentor and starting off from there. If you'd rather start working without one,
 please follow the guidelines in this file.

If you have improvements to OpenSesame, send us your pull requests! For those
just getting started, Github has a
[how to](https://help.github.com/articles/using-pull-requests/).

OpenSesame team members will be assigned to review your pull requests. Once the
pull requests are approved by two team members and pass continuous integration
checks, an OpenSesame member will squash and merge your pull request with the 

> master branch. After approval and merging, the site will be redeployed on a 
> monthly basis. This is done to ensure that we get a good amount of progress 
> onto the website before new versions are rolled out.

Reviewer 2: ```suggestion
master branch. The site will be redeployed on a monthly basis, so it may take a while
for your changes to be deployed. Currently we deploy the site manually, which takes a while.
To save us time we want to ensure that we get a good amount of progress 
on the website before new versions are rolled out. If someone would like to automate
our deployment process we would really appreciate it.
```

Reviewer 2: The best practice is to automate deployment and then deploy as often as possible (often multiple times per day). I'm assuming we don't want to deploy multiple times per day because we don't have an automated deployment system set up yet.

Author: I really like this suggestion!

If you want to contribute, start working through the OpenSesame codebase,
navigate to the
[Github "issues" tab](https://github.com/googleinterns/open-sesame/issues) and start
looking through interesting issues. If you are not sure of where to start, then
start by trying one of the smaller/easier issues here i.e.
[issues with the "good first issue" label](https://github.com/googleinterns/open-sesame/labels/good%20first%20issue)
and then take a look at the
[issues with the "contributions welcome" label](https://github.com/googleinterns/open-sesame/labels/stat%3Acontributions%20welcome).
 If you want to get acquainted with the code first, It might be useful to
start by looking into our documentation and proposing/fixing some 
[issues with the "documentation" label](https://github.com/googleinterns/open-sesame/issues?q=is%3Aissue+label%3Adocumentation).
These are issues that we believe are particularly well suited for outside
contributions, often because we probably won't get to them right now and it is 

> easier to find holes in documentation if you are new to the codebase. 

Reviewer 2: ```suggestion
easier to find holes in the documentation if you are new to the codebase. 
```
